src/solver/impls/cvode/cvode.cxx

```diff
@@ -650,33 +739,34 @@ static int cvode_jac(N_Vector v, N_Vector Jv, realtype t, N_Vector y, N_Vector U
 }
 
 /**************************************************************************
- * vector abstol functions
+ * CVODE vector option functions
  **************************************************************************/
 
-void CvodeSolver::set_abstol_values(BoutReal* abstolvec_data,
-                                    std::vector<BoutReal>& f2dtols,
-                                    std::vector<BoutReal>& f3dtols) {
-  int p = 0; // Counter for location in abstolvec_data array
+void CvodeSolver::set_vector_option_values(BoutReal* option_data,
+                                           std::vector<BoutReal>& f2dtols,
+                                           std::vector<BoutReal>& f3dtols) {
+  int p = 0; // Counter for location in option_data array
 
   // All boundaries
   for (const auto& i2d : bout::globals::mesh->getRegion2D("RGN_BNDRY")) {
-    loop_abstol_values_op(i2d, abstolvec_data, p, f2dtols, f3dtols, true);
+    loop_vector_option_values_op(i2d, option_data, p, f2dtols, f3dtols, true);
   }
   // Bulk of points
   for (const auto& i2d : bout::globals::mesh->getRegion2D("RGN_NOBNDRY")) {
-    loop_abstol_values_op(i2d, abstolvec_data, p, f2dtols, f3dtols, false);
+    loop_vector_option_values_op(i2d, option_data, p, f2dtols, f3dtols, false);
   }
 }
 
-void CvodeSolver::loop_abstol_values_op(Ind2D UNUSED(i2d), BoutReal* abstolvec_data,
-                                        int& p, std::vector<BoutReal>& f2dtols,
-                                        std::vector<BoutReal>& f3dtols, bool bndry) {
+void CvodeSolver::loop_vector_option_values_op(Ind2D UNUSED(i2d), BoutReal* option_data,
+                                               int& p, std::vector<BoutReal>& f2dtols,
+                                               std::vector<BoutReal>& f3dtols, bool bndry)
+{
   // Loop over 2D variables
   for (std::vector<BoutReal>::size_type i = 0; i < f2dtols.size(); i++) {
     if (bndry && !f2d[i].evolve_bndry) {
       continue;
     }
-    abstolvec_data[p] = f2dtols[i];
+    option_data[p] = f2dtols[i];
```

Comment:
Don't understand this error - I don't think there's any pointer arithmetic here, just accessing a C-array with an `int` index...

Reply:
The C-array indexing is basically sugar for pointer arithmetic: the compiler turns it into `*(option_data + p) = f2dtols[i]` (and this is why `array[0] == 0[array]` works even though it looks like bad magic). This is a bit more obvious looking at the function signature, which takes a `BoutReal*` rather than an array type.

There isn't a great deal we can do to do this properly, given that we don't own `N_Vector` (and also that it's pretty much an opaque pointer). In C++20, we could take a `std::span` here instead of `BoutReal*`.

I think we _could_ use `N_VMake_Parallel` to create `abstolvec` with our own data (e.g. `Array`), and then pass that directly to `set_abstol_values`.

(Incidentally, I'm pretty sure we could do something like that to avoid copying the derivatives in and out of the solver's data structures if we could somehow allocate a big block of consecutive memory for the derivatives in the first place)